[WIP] Linux: set broadcast flag by default for ipvlan interfaces.

ido opened the pull request:
The patch uses ethtool's `SIOCETHTOOL` `ioctl` to obtain the driver name of the interface, and turns on the broadcast flag by default when it detects an `ipvlan` interface.  This does not yet differentiate between L2, L3, or L3S mode.

@rsmarples we can use this `ethtool` method to detect most other device types, like `bridge` and `tuntap`, without reading `/sys`.  I might refactor those after this is cleaned up.

TODO:
- [ ] Modify DHCP IAID so that `ipvlan` interfaces do not generate an IAID conflict: `ipvlan0: IAID conflicts with one assigned to eth0`
- [x] Clean up patch before merging.

Fixes #32

ido committed:
Linux: use broadcast flag with ipvlan interfaces by default.

Linux ipvlan interfaces share a MAC address with their siblings and
parent physical interface.  Before they are assigned an IP address,
these virtual interfaces do not receive DHCP OFFER unicast messages
because the ipvlan driver does not know to pass them to the virtual
interface yet by IP.  This chicken-and-egg problem is resolved with
two changes:

In this patch, we set the broadcast flag for an interface if it
belongs to the ipvlan driver, as detected via SIOCETHTOOL ETHTOOL_GDRVINFO.
(closes #32)

A forthcoming patch will automatically modify the DHCP IAID for
ipvlan interfaces so that they do not conflict with the parent
(lower/physical) interface IAID.  For now, dhcpcd will display a warning
log message when conflicting IAID (same MAC address) interfaces are active.

(A minor grammar correction is included free of charge.)

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+	struct ifreq ifr;
+	struct ethtool_drvinfo drvinfo;
+	int fd;
+	if (!ifname || !*ifname)
```
dhcpcd strives to follow NetBSD KNF style.

As such a newline after variable decs is required.
We prefer explicit NULL checks, but it's not mandatory.
Also, NULL is not the NULL character.
Lastly, if returning an error, set errno.

```
if (ifname == NULL || *ifname == '\0') {
    errno = EINVAL;
```

Infact checking if the first character in ifname is the NULL character is redunant as the ioctl will return a suitable
error for us.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+static int
+if_get_driver(const char *ifname, char *driver, const size_t maxlen)
+{
+	struct ifreq ifr;
```
You can avoid the memset below by setting a member to 0.
`struct ifreq ifr = { .ifr_flags = 0 };`

ido replied:
Doing as you suggest, but wouldn't advise it in general because padding would not be zeroed, which when applied to data that passes over a socket, can leak information. (More relevant in the privileged to unprivileged direction.)

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+	memset(&ifr, 0, sizeof(ifr));
+	strlcpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+	drvinfo.cmd = ETHTOOL_GDRVINFO;
+	ifr.ifr_data = (void *)&drvinfo;
```
ifr_data is void anyway, so the cast shouldn't be required.

ido replied:
It's `__caddr_t` (`char *`) on my system, which generates a compiler warning without this cast.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+	strlcpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+	drvinfo.cmd = ETHTOOL_GDRVINFO;
+	ifr.ifr_data = (void *)&drvinfo;
+	if (ioctl(fd, SIOCETHTOOL, &ifr) < 0) {
```
ioctl on Linux always returns -1 on error.
Some ioctls can return negative values as success, so an explicit -1 test is generally better.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+	ifr.ifr_data = (void *)&drvinfo;
+	if (ioctl(fd, SIOCETHTOOL, &ifr) < 0) {
+		const int err = errno;
+		logerr("if_get_driver: failed ethtool ioctl on interface %s: %s",
```
logerr() will append strerror(errno) and preserve errno.
logerrx() does not.

`logerr("%s: SIOCETHTOOL ifname=%s", __func__, ifname);`

might be better.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+	}
+	close(fd);
+	return strlcpy(driver, drvinfo.driver,
+		MIN(sizeof(drvinfo.driver), maxlen));
```
MIN here is redundant because the size of the drvinfo.driver property is irrelevant to the size of the driver variable.

`strlcpy(driver, drvinfo.driver, maxlen);`

is fine.

ido replied:
Assumes drvinfo.driver is properly null-terminated, but fair point.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
 
+/* @maxlen should be greater than or equal to 32; see ethtool_drvinfo. */
+static int
+if_get_driver(const char *ifname, char *driver, const size_t maxlen)
```
maxlen is a poor name as more parameters could be added. driverlen would be better as it's more descriptive as to it's purpose.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+static bool
+if_ipvlan(const char *ifname)
+{
+	char driver[32]; /* sizeof(ethtool_drvinfo.driver) = 32 */
```
Should use `IF_NAMESIZE` rather than 32 as `IF_NAMESIZE` is defined by the system headers.
A newline after the variable decs would be nice.

ido replied:
I did it this way to minimize the amount of times SIOCETHTOOL is called in case we end up using this method to detect many other driver types: bridge, tuntap, macvlan, etc. are all detectable this way instead of using the sysfs method currently in use.  (I think SIOCETHTOOL might even be preferred to sysfs for driver detection.)

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+{
+	char driver[32]; /* sizeof(ethtool_drvinfo.driver) = 32 */
+	if (if_get_driver(ifname, driver, sizeof(driver)) < 0) {
+		logerr("if_ipvlan: failed to get driver name %s.", ifname);
```
You can use __func__ with logerr like so
`logerr("%s: failed to get driver name %s", __func__, ifname);`
Generally debug lines don't end with a full stop.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
+		logerr("if_ipvlan: failed to get driver name %s.", ifname);
+		return false;
+	}
+	return (strncmp(driver, "ipvlan", sizeof(driver)) == 0);
```
the extra braces around strncmp are superfluous.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
 	int n;
 
+	/* Set broadcast flag for ipvlan interfaces. */
+	if (if_ipvlan(ifp->name)) {
```
Braces for a single statement are superfluous.
No biggie though, that's personal preference.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,60 @@ if_init(struct interface *ifp)
 
+	/* Set broadcast flag for ipvlan interfaces. */
+	if (if_ipvlan(ifp->name)) {
+		ifp->options->options |= DHCPCD_BROADCAST;
```
The only downside of setting this here rather than in dhcpcd.c is that if any other OS gains ipvlan functionality then this logic might be forgotten about.

That's why I created stub functions in if-bsd.c and if-sun.c that just return false.

ido replied:
Added an `XXX` to remind myself of this, but AFAIK `ipvlan` is unique to Linux.  (There are other OSes with L3IPVLAN interfaces, but they behave differently.)

rsmarples reviewed (changes requested):
Mainly comments on style.
Otherwise looks good!

deliciouslytyped commented:
Please remember to change the documentation as well!

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,70 @@ if_init(struct interface *ifp)
+		logerr("%s: socket ifname=%s", __func__, ifname);
+		goto eexit;
+	}
+	memset(&ifr, 0, sizeof(ifr));
```
This can now be removed.
BTW, you can clarify the structure is correctly zeroed by using Valgrind or compiling with the address or memory sanitizer provided by your compiler.
Although you might need to disable privsep, entirely or just the sandbox + chroot + privdrop to get it to work.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,70 @@ if_init(struct interface *ifp)
+if_get_driver(const char *ifname, char *driver, const size_t driverlen)
+{
+	struct ifreq ifr = { .ifr_flags = 0 };
+	struct ethtool_drvinfo drvinfo;
```
This structure needs initialisation or it may pass unintended data to the ioctl.

rsmarples replied:
A quick glance at the structure shows it should have no padding on all arches so it could be initialised in the same way.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,70 @@ if_init(struct interface *ifp)
+static bool
+if_cmp_driver(const char *ifname, const char *driver)
+{
+	char ifdriver[32]; /* see ethtool_drvinfo.driver declaration */
```
Does this work? (untested)

`char ifdriver[sizeof(((struct ethtool_drvinfo *)0)->driver)]`

ido replied:
Yes, that's FIELD_SIZEOF in the Linux kernel source, I'll add it to `common.h` as part of this patch. 👍

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,70 @@ if_init(struct interface *ifp)
+	int fd;
+	size_t n = 0;
+
+	fd = socket(AF_INET, SOCK_STREAM, 0);
```
This may not work in privsep.
Please use if_ioctl, see here for an example:
https://github.com/NetworkConfiguration/dhcpcd/blob/master/src/if-linux.c#L505

rsmarples replied:
Actually that's not enough.
I think we need to move `if_indirect_ioctl` from `if-bsd.c`  to `if.c`
https://github.com/NetworkConfiguration/dhcpcd/blob/master/src/if.c#L149

Also move `PS_IOCTLINDIRECT` and friends from `privsep-bsd.c` to`privsep-root.c`
That's quite an invasive change though.

ido replied:
IIUC, SIOCETHTOOL with ETHTOOL_GDRVINFO should not require root on most systems, and that's the only kernel we need to support this on (for now).  (i.e. `ethtool -i $INTERFACE` should work for any user.)

Currently, `if_vlanid(...)` uses `ifp->ctx->pf_inet_fd` directly.  I could do that instead of creating my own socket, would that be enough for now if we're only supporting `ipvlan` on Linux for this patch?

rsmarples replied:
> Currently, `if_vlanid(...)` uses `ifp->ctx->pf_inet_fd` directly. I could do that instead of creating my own socket, would that be enough for now if we're only supporting `ipvlan` on Linux for this patch? That simplifies things quite a bit.

Yes, that's fine.

ido commented:
I squashed the commits into one patch.  So far, this sets broadcast.  I might have a solution for setting unique IAIDs soon.

rsmarples commented on src/if-linux.c:
```diff
@@ -309,12 +310,55 @@ if_init(struct interface *ifp)
+	strlcpy(ifr.ifr_name, ifp->name, sizeof(ifr.ifr_name));
+	if (ioctl(ifp->ctx->pf_inet_fd, SIOCETHTOOL, &ifr) != 0) {
+		logerr("%s: SIOCETHTOOL ifname=%s", __func__, ifp->name);
+		return 0; /* 0 means error or indeterminate driver name */
```
I would change the function return type to ssize_t and return -1 here to distinguish between an ioctl error and zero length driver name.

ido committed:
Linux: change return value of if_get_driver to ssize_t.

Differentiate between ioctl error and zero-length driver name in if_get_driver.

ido committed:
Linux: fix sign-conversion compiler warning.
